src/state/base.py

```diff
@@ -0,0 +1,29 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""gateway-api-integrator charm state constructor."""
+import dataclasses
+import typing
+
+from .config import CharmConfig
+from .gateway import GatewayResourceDefinition
+from .secret import SecretResourceDefinition
+from .tls import TLSInformation
+
+Components = typing.TypeVar(
+    "Components", GatewayResourceDefinition, CharmConfig, SecretResourceDefinition, TLSInformation
+)
+
+
+class State:  # pylint: disable=too-few-public-methods
```

Comment:
Just out of curiosity, would this wrapper be safe enough to catch the errors statically? For instance, when we forget to pass in an argument, can it catch it statically? The reason I'm asking is because then the checks would need to delegate the responsibility to integration tests which take longer than lint/unit tests.

Reply:
I also subclassed `State` (now `ResourceDefinition`) and override the `__init__` method to explicitly require the necessary components